File: python/binary_search/LC_744_FindSmallestLetterGreaterThanTarget.py
# https://leetcode.com/problems/find-smallest-letter-greater-than-target
from typing import List
#  

class Solution:
    def nextGreatestLetter(self, letters: List[str], target: str) -> str:
        left , right = 0, len(letters)-1
        while left <= right:
            mid = left + (right - left) // 2 
            if target >= letters[mid]:
                left = mid + 1
            else:
                right = mid -1
        return letters[left % len(letters)]
    
sol = Solution()
print(sol.nextGreatestLetter(letters = ["c","f","j"], target = "a"))
print(sol.nextGreatestLetter(letters = ["c","f","j"], target = "c"))

File: java/src/binary_search/LC_744_FindSmallestLetterGreaterThanTarget.java
package binary_search;
// https://leetcode.com/problems/find-smallest-letter-greater-than-target

public class LC_744_FindSmallestLetterGreaterThanTarget {

    public char nextGreatestLetter(char[] letters, char target) {
        int left = 0, right = letters.length;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (target >= letters[mid])
                left = mid + 1;
            else
                right = mid - 1;
        }
        return letters[left % letters.length];
    }

    public static void main(String[] args) {
        LC_744_FindSmallestLetterGreaterThanTarget greaterThanTarget = new LC_744_FindSmallestLetterGreaterThanTarget();
        System.out.println(greaterThanTarget.nextGreatestLetter(new char[]{'c', 'f', 'g'}, 'a'));
    }
}